### PR TITLE
chore: add kongintegration test flavor and cover UpdateStrategyInMemory as a reference

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

Adds an integration test for `sendconfig.UpdateStrategyInMemory` to verify we correctly integrate with Kong's flattened errors.

It also creates a separate "kongintegration" test flavor (name to be discussed, I'm open to suggestions). I think we could benefit from having more coverage in this kind of tests instead of always adding a regular integration or E2E case. They all could be independent and run in parallel with no problems. We could test various Kong settings in them as every test spawns its own Kong instance. Their only dependency is to have a running docker daemon.

**Which issue this PR fixes**:

Follow up of https://github.com/Kong/kubernetes-ingress-controller/pull/4790.

### CODE REVIEW

File: internal/dataplane/sendconfig/inmemory_integration_test.go
Comment (pmalek) [inline internal/dataplane/sendconfig/inmemory_integration_test.go:1]: I like this idea 👍 

We'd definitely need to think about how to structure this:

- do we join this with our current integration tests? (rather not)
- if not, then where do we place it and how do we run those? Perhaps that would warrant a new directory under `test/` and a new test target

File: internal/dataplane/sendconfig/inmemory.go [hunk 1]
```diff
@@ -35,20 +34,17 @@ type UpdateStrategyInMemory struct {
...
 	configConverter ContentToDBLessConfigConverter
 	log             logrus.FieldLogger
-	version         semver.Version
 }
 
 func NewUpdateStrategyInMemory(
 	configService ConfigService,
 	configConverter ContentToDBLessConfigConverter,
 	log logrus.FieldLogger,
-	version semver.Version,
```
Comment (pmalek) [inline internal/dataplane/sendconfig/inmemory.go:45]: If this is not needed anymore then we can get rid of this in a separate pr 👍

**Comment by programmer04** [comment]
Target `test.expression_router` can be collapsed into what you're proposing here, ofc in a separate PR

File: internal/dataplane/sendconfig/inmemory.go (hunk 1 above)
Comment (pmalek) [reply internal/dataplane/sendconfig/inmemory.go]: This is done via #4804

**Comment by czeslavo** [comment]
> Target `test.expression_router` can be collapsed into what you're proposing here, ofc in a separate PR

Yeah, I gave it a shot and it works: https://github.com/Kong/kubernetes-ingress-controller/pull/4806 

I also tried verifying Parser's golden test outputs in this kind of test: https://github.com/Kong/kubernetes-ingress-controller/pull/4805

File: test/kongintegration/inmemory_update_strategy_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		},
+	}
+	expectedResourceError := sendconfig.ResourceError{
+		Name:       "test-service",
+		Namespace:  "default",
+		Kind:       "Service",
+		UID:        "a3b8afcc-9f19-42e4-aa8f-5866168c2ad3",
+		APIVersion: "v1",
+		Problems: map[string]string{
+			"":     "failed conditional validation given value of field 'protocol'",
```
Comment (pmalek) [inline test/kongintegration/inmemory_update_strategy_test.go:76]: Weird that this doesn't have a key 🤔

Comment (czeslavo) [reply test/kongintegration/inmemory_update_strategy_test.go:76]: Yeah, as I understand it's a way to say "It's a problem that applies to a whole entity, not a single field". It's covered in [unit tests](https://github.com/Kong/kubernetes-ingress-controller/blob/ac90e1e2e674a3f369adc6152bf9304eb58d24b9/internal/dataplane/sendconfig/error_handling_test.go#L38) as expected behavior.

Despite that fact, I think it would make sense to fix it, because currently it generates a weird message in the Kubernetes event, e.g.:

```
invalid : failed conditional validation given value of field 'protocol'"
```

Posted a PR with fix for that: https://github.com/Kong/kubernetes-ingress-controller/pull/4813

File: test/kongintegration/inmemory_update_strategy_test.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			"path": "value must be null",
+		},
+	}
+
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
```
[the remark this answers is not in the document]
Comment (czeslavo) [reply test/kongintegration/inmemory_update_strategy_test.go:81]: With a polling period higher than milliseconds it should work fine as that would be hard to trigger a race condition in such circumstances. 

Anyway, I think it makes sense to postpone using it until https://github.com/stretchr/testify/pull/1395 is merged and released. (I've just got another round of review after quite some time, so there's hope we'll be able to finally close it 🤞).

File: .github/workflows/_kongintegration_tests.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+      - name: setup golang
+        uses: actions/setup-go@v4
+        with:
+          go-version-file: go.mod
+
+      - name: set kong version
+        run: |
+          echo "TEST_KONG_IMAGE=kong" >> $GITHUB_ENV
+          echo "TEST_KONG_TAG=$(cat .github/test_dependencies.yaml | yq -ojson -r '.kongintegration.kong-oss')" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (czeslavo) [reply .github/workflows/_kongintegration_tests.yaml:23]: Yeah, we definitely should allow testing against the enterprise version somehow. Running a parallel job in the workflow might be a good way to do that. We'll probably have some tests that should be run only against EE (where we test EE features) and we should account for that in the tests themselves by inspecting the image we've passed in env.

Added an issue to track this: https://github.com/Kong/kubernetes-ingress-controller/issues/4815

**Review by pmalek — APPROVED** [review]

File: test/kongintegration/inmemory_update_strategy_test.go (hunk 2 above)
Comment (pmalek) [reply test/kongintegration/inmemory_update_strategy_test.go:76]: 😎

File: test/kongintegration/kongintegration.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
+// Package kongintegration contains integration tests that require a Kong instance to be running, but
+// do not require a Kubernetes cluster nor full Kong Ingress Controller deployment.
+//
+// All tests that verify KIC's individual components compatibility with Kong should be placed here.
```
[the remark this answers is not in the document]
Comment (czeslavo) [reply test/kongintegration/kongintegration.go:4]: Good idea. Done this + renamed `kongintegration.go` into just `kong.go`.

File: test/kongintegration/kong.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+// kongImageUnderTest returns the Kong image to be used for integration tests. If both TEST_KONG_IMAGE and
+// TEST_KONG_TAG are set, it will return the image and tag specified by them. Otherwise, it will return
+// the default image (kong:latest).
+func kongImageUnderTest() string {
+	if testenv.KongImage() != "" && testenv.KongTag() != "" {
+		return fmt.Sprintf("%s:%s", testenv.KongImage(), testenv.KongTag())
+	}
+
+	return "kong:latest"
```
Comment (pmalek) [inline test/kongintegration/kong.go:54]: This could be shared with the integration tests.

Let's do this in a separate PR to not duplicate code unnecessarily.

File: .github/workflows/checks.yaml [hunk 7]
```diff
@@ -91,6 +91,13 @@ jobs:
     uses: ./.github/workflows/_envtest_tests.yaml
     secrets: inherit
 
+  kongintegration-tests:
```
Comment (pmalek) [inline .github/workflows/checks.yaml:94]: Aren't we adding this to `passed` `needs`?

Comment (czeslavo) [reply .github/workflows/checks.yaml:94]: Good catch :) PTAL

**Review by pmalek — APPROVED** [review]
